File: src/lib/queue.ts
import { invoke } from "@tauri-apps/api/core";
import type { FormatPreference, ReformatOutcome } from "@/lib/enqueue";

export type QueueStatus = "queued" | "downloading" | "completed" | "failed" | "cancelled";

export interface QueueEntry {
  id: number;
  video_id: string;
  title: string;
  itag: number;
  quality_label: string | null;
  output_path: string;
  convert_to_mp3: boolean;
  status: QueueStatus;
  error_message: string | null;
  created_at: number;
}

/**
 * Stand-in itag on an MP3 entry whose source stream yt-dlp picks at download
 * time, because nothing in the video's format list matched (mirrors
 * `core::stream::AUTO_AUDIO_ITAG`). Not a real itag, so it isn't worth
 * showing the user a number for.
 */
export const AUTO_AUDIO_ITAG = 0;

/** What an entry's format line reads as. */
export function formatDescription(entry: QueueEntry): string {
  const source = entry.itag === AUTO_AUDIO_ITAG ? "best audio" : `itag ${entry.itag}`;
  return entry.convert_to_mp3
    ? `mp3 (from ${source})`
    : `${entry.quality_label ?? "audio"} (${source})`;
}

export interface AddToQueueParams {
  videoId: string;
  title: string;
  itag: number;
  qualityLabel: string | null;
  outputPath: string;
  convertToMp3?: boolean;
  [key: string]: unknown;
}

export const addToQueue = (params: AddToQueueParams) => invoke<QueueEntry>("add_to_queue", params);

export const listQueue = () => invoke<QueueEntry[]>("list_queue");

export const removeFromQueue = (queueId: number) =>
  invoke<void>("remove_from_queue", { queueId });

export const clearQueue = () => invoke<void>("clear_queue");

export interface SetQueueEntryFormatParams {
  queueId: number;
  itag: number;
  qualityLabel: string | null;
  convertToMp3: boolean;
  [key: string]: unknown;
}

/** Repoints one entry at an exact itag the user picked for that video. */
export const setQueueEntryFormat = (params: SetQueueEntryFormatParams) =>
  invoke<QueueEntry>("set_queue_entry_format", params);

export interface SetQueueEntriesQualityParams {
  queueIds: number[];
  preference: FormatPreference;
  [key: string]: unknown;
}

/**
 * Re-resolves several entries against one quality preference. A preference
 * rather than an itag, because a multi-select spans videos whose available
 * itags differ — the backend resolves each video individually.
 */
export const setQueueEntriesQuality = (params: SetQueueEntriesQualityParams) =>
  invoke<ReformatOutcome>("set_queue_entries_quality", params);

File: src-tauri/src/commands/queue.rs
//! Thin Tauri command handlers for the download queue. All persistence
//! logic lives in `downloadhub_core::queue`; this module just wires it up.
//!
//! Command names (`add_to_queue`, `list_queue`) match the MCP tool names
//! planned for Phase 3 (see `CLAUDE.md`), so the MCP server can expose the
//! same underlying `downloadhub_core::queue` operations without renaming.

use crate::state::AppState;
use downloadhub_core::enqueue::{self, ReformatOutcome};
use downloadhub_core::queue::{NewQueueEntry, QueueEntry, QueueStatus};
use downloadhub_core::stream::{FormatPreference, ResolvedFormat};
use tauri::State;

#[tauri::command]
pub async fn add_to_queue(
    video_id: String,
    title: String,
    itag: u32,
    quality_label: Option<String>,
    output_path: String,
    // `Option` so existing callers that don't send the field keep working
    // (Tauri rejects a missing required argument).
    convert_to_mp3: Option<bool>,
    state: State<'_, AppState>,
) -> Result<QueueEntry, String> {
    state
        .queue_store()?
        .add_entry(NewQueueEntry {
            video_id,
            title,
            itag,
            quality_label,
            output_path,
            convert_to_mp3: convert_to_mp3.unwrap_or(false),
        })
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_queue(state: State<'_, AppState>) -> Result<Vec<QueueEntry>, String> {
    state
        .queue_store()?
        .list_entries()
        .await
        .map_err(|e| e.to_string())
}

/// Repoints one entry at an exact format the user picked from that video's
/// own format list, resetting it to `Queued`. The precise counterpart to
/// [`set_queue_entries_quality`], which can only work in preferences
/// because a multi-select spans videos with different itags.
#[tauri::command]
pub async fn set_queue_entry_format(
    queue_id: i64,
    itag: u32,
    quality_label: Option<String>,
    convert_to_mp3: bool,
    state: State<'_, AppState>,
) -> Result<QueueEntry, String> {
    state.ensure_no_batch_running()?;
    let store = state.queue_store()?;

    // Changing the format under a running download would leave the task
    // writing the old stream to a path derived from the new one.
    match store.get_entry(queue_id).await.map_err(|e| e.to_string())? {
        Some(entry) if entry.status == QueueStatus::Downloading => {
            return Err("This entry is downloading; cancel it before changing its format.".into());
        }
        Some(_) => {}
        None => return Err("That queue entry no longer exists.".to_string()),
    }

    store
        .set_format(
            queue_id,
            &ResolvedFormat {
                itag,
                quality_label,
                convert_to_mp3,
            },
        )
        .await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "That queue entry no longer exists.".to_string())
}

/// Re-resolves several entries against a single quality preference (the
/// "update all selected" path). Each video's formats are looked up
/// individually, so this makes one network round-trip per entry; entries
/// that fail to resolve are reported in `skipped` rather than sinking the
/// whole update.
#[tauri::command]
pub async fn set_queue_entries_quality(
    queue_ids: Vec<i64>,
    preference: FormatPreference,
    state: State<'_, AppState>,
) -> Result<ReformatOutcome, String> {
    state.ensure_no_batch_running()?;
    enqueue::reformat_entries(
        &state.stream_client,
        state.queue_store()?,
        &queue_ids,
        preference,
    )
    .await
    .map_err(|e| e.to_string())
}

/// Removes an entry from the queue. If a download is currently running for
/// it, that task is aborted first — otherwise it would keep writing to a
/// file its queue record no longer exists for.
#[tauri::command]
pub async fn remove_from_queue(queue_id: i64, state: State<'_, AppState>) -> Result<(), String> {
    state.ensure_no_batch_running()?;

    if let Some(handle) = state
        .running_downloads
        .lock()
        .expect("running downloads mutex poisoned")
        .remove(&queue_id)
    {
        handle.abort();
    }

    state
        .queue_store()?
        .delete_entry(queue_id)
        .await
        .map_err(|e| e.to_string())
}
